libowfat/io/iom_wait.3
.TH iom_wait 3
.SH NAME
iom_wait \- wait for event from I/O multiplexer
.SH SYNTAX
.B #include <io.h>

int \fBiom_wait\fP(iomux_t* c,
                   int64* fd, unsigned int* events,
		   unsigned long timeout);
.SH DESCRIPTION
iom_wait will wait for events registered to the I/O multiplexer with
\fIiom_add\fR. It will wait \fItimeout\fR milliseconds.

If during that time any of the registered events occur, \fIiom_wait\fR
will set \fIfd\fR to the file descriptor the event happened on, and
\fIevents\fR to the sum of IOM_READ, IOM_WRITE and IOM_ERROR, depending
on what event actually happened, and return 1.

If nothing happens during that time, it will return 0 and leave \fIfd\fR
and \fIevents\fR alone.

Note that the event registration is removed from the iomux_t context if
it occurs. You will have to call \fIiom_wait\fR again after you handled
the event, if you are still interested in it.

Closing a file descriptor with registered events will discard the event
registration.

.SH "LINKING"
You may have to add \fI-lpthread\fR to the command line in the linking
step.

.SH "RETURN VALUE"
iom_wait returns 1 on success, 0 if there was a timeout, and -1 on
error, setting errno. If \fIiom_abort\fR was called on the I/O
multiplexer context, it will return -2.
.SH "SEE ALSO"
iom_init, iom_add, iom_abort


libowfat/CHANGES
0.32:
  remove OpenBSD #warning (obsd maintainer says no longer needed)
  move headers to <libowfat/> upon install
  fix fmt_ip6 (Erwin Hoffmann)
  add MSG_ZEROCOPY support (only used for buffers >8k)
  use writev in buffer_put for a slight perf improvement
  add experimental iom API for multithreaded I/O multiplexing (in io.h)

0.31:
  special case buffer_get_token with token length 1 through memccpy (almost 4x speedup)
  test for pread and use it instead of lseek+read in cdb and io_mmapwritefile
  add mmap_readat (like mmap_read but open by fd, not by name)
  add fmt_iso8601 and scan_iso8601 (time_t -> "2014-05-27T19:22:16Z")
  add fmt_netstring and scan_netstring (http://cr.yp.to/proto/netstrings.txt)
  add (fmt|scan)_varint and (fmt|scan)_pb_* (binary marshaling ala protocol buffers)

0.30:
  add compiletimeassert.h
  add haveuint128.h
  byte.h now defines UNALIGNED_ACCESS_OK on x86
  add scan_ulongn, scan_8longn, scan_xlongn, scan_longn, scan_netstring
  change buffer to have a destructor function pointer
  SECURITY: fmt_strn would write one byte too many (returned right length though, so usually not a problem as that byte would be overwritten with \0 by the caller anyway)
  fmt_pad and fmt_fill fail more gracefully when srclen > maxlen
  You can now say $ make WERROR=-Werror (compiling t.c may fail but that's strictly speaking not part of the library)
  scan_html now decodes HTML entities based on the actual w3c list (from entities.json, say gmake update to fetch the current version)
  added fmt_escapechar* to fmt.h (implement various escaping mechanisms also found in textcode but for a single char not a whole string, and they always escape, not just when they think it's needed)
  scan_ushort was supposed to abort early and return 5 when attempting to parse "65536", because the result does not fit.  It did not.  Now it does.
  scan_*long, scan_*int, scan_*short now properly abort if the number would not fit
  SECURITY: check for integer overflow in stralloc_ready
  switch io_fds from array to newly implemented (hopefully thread-safe) iarray
  switch epoll from level triggering to edge triggering
  introduce io_eagain_read and io_eagain_write (discontinue using io_eagain plz)
  fix buffer_get
  add fmt_html_tagarg, fmt_xml, scan_html_tagarg
  add socket_fastopen, socket_fastopen_connect4, socket_fastopen_connect6 and socket_quickack
  on Linux, in iob_send, if MSG_MORE is defined, and the request type
    lends itself to it, used sendto/sendmsg with MSG_MORE instead of
    TCP_CORK (saves two syscalls)
  add io_fd_flags so the caller can tell io_fd whether the socket is blocking
    (saves one fcntl syscall)
  more constness for stralloc and buffer
  mmap_read/mmap_shared on zero length files no longer fail but return a
    zero length buffer
  if SOCK_NONBLOCK is defined, use it instead of socket+fcntl
  ... but if errno==EINVAL still fall back to socket+fcntl (Robert Henney)
  SECURITY: fix botched integer overflow handling logic in stralloc_ready (Giorgio)
  add critbit

0.29:
  save 8 bytes in taia.h for 64-bit systems
  add buffer_tosa (buffer writing to auto-growing stralloc)
  add iarray
  support SCTP in addition to TCP
  fix io_receivefd so the incoming buffer is only 1 byte; io_passfd
    sends only one byte, so we might receive (and discard) other data if
    we try to read more (Stefan Bühler, from the lightttpd project)
  add io_fd_canwrite (like io_fd but assume the fd is writable)
  save a few syscalls here and there
  add awesome hack in isset.h (comex, Linus Torvalds)
  add byte_equal_notimingattack

0.28:
  add uint64 pack and unpack routines
  fix subtle typo in sub_of (David Sirovsky)
  work around gcc bugs

0.27:
  add fmt_strm
  add iob_addbuf_munmap
  add socket_deferaccept
  catch one more case in umult64 (Tomi Jylhä-Ollila)

0.26:
  fix really pathological case where io_timeouted would never
    start over from the beginning because always new accept()ed
    connections came in and got newer, higher descriptors since the last
    io_timeouted loop.  (Dirk Engling)
  add some int overflow check macros to rangecheck.h
  fmt_ip6 compresses at best spot, not at first spot (Nikola Vladov)
  use inttypes.h to declare ints in uint*.h
  escape more in fmt_ldapescape
  try to catch malicious input in textcode fmt_* functions
  fmt_xlonglong was utterly broken (Johannes Vetter)

0.25:
  array_allocate no longer truncates the array
  array_get now not only checks whether the element fits in the
    allocates space, it also needs to be in the initialized space.
  add -D_REENTRANT to CFLAGS so libowfat can be used in multi-threaded
    programs
  further Windoze support (test/io5.c works, gatling still doesn't)
    This is just to get gatling to work, I may remove it again after
    that.
  implement Nikola's idea to remove limit number of strings in errmsg
  add taia_half
  add cdb
  add rangecheck.h
  add io_block
  make socket_(tc|ud)p[46] actually return non-blocking sockets as
    documented (Richard Lyons)
  fix for NetBSD 3.0 (Their /sys/types.h actually does not define u_char
    unless you #define _NETBSD_SOURCE, and then their CMSG_* macros use
    it.  This is horrible even by OpenBSD standards)
  remove support for ip6.int (it's no longer delegated)
  add asm versions of imult64 and umult64 for x86_64
    (22 cycles -> 12 cycles on my Athlon 64)
  use size_t and ssize_t instead of unsigned long et al
  add iob_bytesleft
  don't leak memory in test/httpd

0.24:
  fix scan_to_sa (Tim Lorenz)
  turns out we can not do file descriptor passing on OpenBSD
    if _XOPEN_SOURCE is defined, which is needed for Solaris.
    OpenBSD sucks.  I checked in a _really_ kludge now.  Please do not
    use OpenBSD if you have a choice.  Or Solaris, for that matter.
  turns out the imult routines (which I never used) were incorrect.
    Noted by Matthew Dempsky
  open_* from open.h now open in large file mode

0.23:
  also recognize EPFNOSUPPORT as EAFNOSUPPORT (groan)

libowfat/io/iom_wait.c
#include "io_internal.h"
#ifdef HAVE_EPOLL
#include <sys/epoll.h>
#endif
#ifdef HAVE_KQUEUE
#include <sys/types.h>
#include <sys/event.h>
#include <sys/time.h>
#endif
#include <errno.h>

int iom_wait(iomux_t* c,int64* s,unsigned int* revents,unsigned long timeout) {
  for (;;) {
    /* If we have an event in the queue, use that one */
    int r;
    if (c->working==-2) return -2;	/* iomux was aborted */
    for (;;) {
      unsigned int f=c->l;
      if (f == c->h)
	break; /* no elements in queue */
      int n=(f+1)%SLOTS;
      if (__sync_bool_compare_and_swap(&c->l,f,n)) {
	/* we got one, and its index is in f */
	*s=c->q[f].fd;
	*revents=c->q[f].events;
      }
      /* collided with another thread, try again */
    }
    /* The queue was empty. If someone else is already calling
    * epoll_wait/kevent, then use the semaphore */
    if (__sync_bool_compare_and_swap(&c->working,0,1)) {
      /* we have the job to fill the struct. */
      int freeslots = (c->h - c->l);
      if (!freeslots) freeslots=SLOTS;

#ifdef HAVE_EPOLL
      struct epoll_event ee[SLOTS];
      int i;
      r=epoll_wait(c->ctx, ee, freeslots, timeout);
      if (r<=0) {
	/* we ran into a timeout, so let someone else take over */
	if (__sync_val_compare_and_swap(&c->working,1,0)==-2) return -2;
#ifdef __dietlibc__
	cnd_broadcast(&c->sem);
#else
	sem_post(&c->sem);
#endif
	return r;
      }
      for (i=0; i<r; ++i) {
	/* convert events */
	int e = ((ee[i].events & (EPOLLIN|EPOLLHUP|EPOLLERR)) ? IOM_READ : 0) |
	        ((ee[i].events & (EPOLLOUT|EPOLLHUP|EPOLLERR)) ? IOM_WRITE : 0) |
	        ((ee[i].events & EPOLLERR) ? IOM_ERROR : 0);
	if (i+1==r) {
	  /* return last event instead of enqueueing it */
	  *s=ee[i].data.fd;
	  *revents=e;
	} else {
	  c->q[c->h].fd=ee[i].data.fd;
	  c->q[c->h].events=e;
	  c->h = (c->h + 1) % SLOTS;
	}
      }
#elif defined(HAVE_KQUEUE)
      struct kevent kev[SLOTS];
      struct timespec ts = { .tv_sec=timeout/1000, .tv_nsec=(timeout%1000)*1000000 };
      int r=kevent(c->ctx, 0, 0, &kev, freeslots, &ts);
      if (r<=0) {
	/* we ran into a timeout, so let someone else take over */
	if (__sync_val_compare_and_swap(&c->working,1,0)==-2) return -2;
#ifdef __dietlibc__
	cnd_broadcast(&c->sem);
#else
	sem_post(&c->sem);
#endif
	return r;
      }
      for (i=0; i<r; ++i) {
	/* convert events */
	int e = (kev[i].filter == EVFILT_READ ? IOM_READ : 0) |
	        (kev[i].filter == EVFILT_WRITE ? IOM_WRITE : 0);
	if (i+1==r) {
	  /* return last event instead of enqueueing it */
	  *s=kev.ident;
	  *revents=e;
	} else {
	  c->q[c->h].fd=kev[i].ident;
	  c->q[c->h].events=e;
	  c->h = (c->h + 1) % SLOTS;
	}
      }
#else
#warning "only epoll and kqueue supported for now"
#endif
      /* We need to signal the other threads.
	 Either there are other events left, or we need one of them to
	 wake up and call epoll_wait/kevent next, because we aren't
	 doing it anymore */
      if (__sync_val_compare_and_swap(&c->working,1,0)==-2) return -2;
#ifdef __dietlibc__
      cnd_signal(&c->sem);
#else
      sem_post(&c->sem);
#endif
      return 1;
    } else {
      /* somebody else has the job to fill the queue */
      struct timespec ts;
      ts.tv_sec = timeout / 1000;
      ts.tv_nsec = (timeout % 1000) * 1000000;
#ifdef __dietlibc__
      r=cnd_timedwait(&c->sem,&c->mtx,&ts);
#else
      r=sem_timedwait(&c->sem,&ts);
#endif
      if (r==-1) {
	if (errno==ETIMEDOUT) return 0;
	return -1;
      }
      /* fall through into next loop iteration */
    }
  }
}
